### ISSUE TITLE
Error if host has more than 8 CPUs with virt guest

### ISSUE THREAD

**Issue by smb1899** [utmapp/UTM#3180]
**Describe the issue**
I have a problem installing windows 11, when I try to start the VM it gives me an error  message (qemu-aarch64-softmmu: Number of SMP CPUs requested (16) exceeds max CPUs supported by machine 'mach-virt' (8))

**Configuration**
* UTM Version: 2.2.4 (36)
* OS Version: macOS Big Sur V. 11.6
* Intel or Apple Silicon? 2,3 GHz 8-Core Intel Core i9

**Debug log**
Running:  -L /private/var/folders/2b/plq7c1w11tdgwzw6mv61pw1m0000gn/T/AppTranslocation/96037664-92D5-49D4-8042-296ED65940FD/d/UTM.app/Contents/Resources/qemu -S -qmp tcp:127.0.0.1:4000,server,nowait -nodefaults -vga none -spice "unix=on,addr=/Users/serajburshan/Library/Group Containers/WDNLXAD4W8.com.utmapp.UTM/14A2D25C-6951-4952-8F51-2A14DB87812C.spice,disable-ticketing=on,image-compression=off,playback-compression=off,streaming-video=off,gl=on" -device virtio-ramfb-gl -cpu cortex-a72 -smp cpus=16,sockets=1,cores=8,threads=2 -machine virt,highmem=off -accel tcg,tb-size=4096 -drive if=pflash,format=raw,unit=0,file=/private/var/folders/2b/plq7c1w11tdgwzw6mv61pw1m0000gn/T/AppTranslocation/96037664-92D5-49D4-8042-296ED65940FD/d/UTM.app/Contents/Resources/qemu/edk2-aarch64-code.fd,readonly=on -drive "if=pflash,format=raw,unit=1,file=/Users/serajburshan/Library/Containers/com.utmapp.UTM/Data/Documents/Windows 11.utm/Images/efi_vars.fd" -boot menu=on -m 16384 -device intel-hda -device hda-duplex -name "Windows 11" -device qemu-xhci,id=usb-bus -device usb-tablet,bus=usb-bus.0 -device usb-mouse,bus=usb-bus.0 -device usb-kbd,bus=usb-bus.0 -device ich9-usb-ehci1,id=usb-controller-0 -device ich9-usb-uhci1,masterbus=usb-controller-0.0,firstport=0,multifunction=on -device ich9-usb-uhci2,masterbus=usb-controller-0.0,firstport=2,multifunction=on -device ich9-usb-uhci3,masterbus=usb-controller-0.0,firstport=4,multifunction=on -chardev spicevmc,name=usbredir,id=usbredirchardev0 -device usb-redir,chardev=usbredirchardev0,id=usbredirdev0,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev1 -device usb-redir,chardev=usbredirchardev1,id=usbredirdev1,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev2 -device usb-redir,chardev=usbredirchardev2,id=usbredirdev2,bus=usb-controller-0.0 -device nvme,drive=drive0,serial=drive0,bootindex=0 -drive "if=none,media=disk,id=drive0,file=/Users/serajburshan/Library/Containers/com.utmapp.UTM/Data/Documents/Windows 11.utm/Images/Windows11_InsiderPreview_Client_ARM64_en-us__22454.VHDX,cache=writethrough" -device usb-storage,drive=drive1,removable=true,bootindex=1 -drive if=none,media=cdrom,id=drive1 -device virtio-net-pci,mac=7E:E2:B0:35:8F:BD,netdev=net0 -netdev vmnet-macos,mode=shared,id=net0 -device virtio-serial -device virtserialport,chardev=vdagent,name=com.redhat.spice.0 -chardev spicevmc,id=vdagent,debug=0,name=vdagent -device virtserialport,chardev=charchannel1,id=channel1,name=org.spice-space.webdav.0 -chardev spiceport,name=org.spice-space.webdav.0,id=charchannel1 -uuid 14A2D25C-6951-4952-8F51-2A14DB87812C -rtc base=localtime
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: Started vmnet interface with configuration:
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: MTU:              1500
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: Max packet size:  1514
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: MAC:              12:75:7f:40:ae:a3
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: DHCP IPv4 start:  192.168.64.1
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: DHCP IPv4 end:    192.168.64.254
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: IPv4 subnet mask: 255.255.255.0
qemu-aarch64-softmmu: -netdev vmnet-macos,mode=shared,id=net0: info: UUID:             4CB43171-F1ED-4FF3-9826-B67E175D7E93
qemu-aarch64-softmmu: Number of SMP CPUs requested (16) exceeds max CPUs supported by machine 'mach-virt' (8)

**Comment by conath** [comment]
What's happening here is a bug in the way UTM configures itself automatically: it takes the number of logical CPU cores your Mac has and tries to give that many to the VM, regardless if the VM system can take it.

To work around this issue, please go to the VM configuration → System → Advanced and type "4" in the CPU Cores field. Then enable "Force Multicore" to improve performance.

![Screen Shot 2021-10-16 at 21 09 13](https://user-images.githubusercontent.com/12073163/137599429-8a960da7-8555-4d57-b376-b4aa9299353c.jpg)

I will leave this Issue open and change the title to reflect the source of the problem.

**Comment by osy** [comment]
Thinking of making two changes. Let me know if there’s any objections:

* New installs and those who never touched the options will have the “Use performance cores by default” option set. This way people won’t have weird performance inbalances.
* If the host cpu cores are > the max for virt, we set it to max (8). Ideally we get this info from QEMU but it doesn’t exist outside of the source code so we’ll have to manually code in a list of popular machines and their max cpu count.

**Comment by conath** [comment]
Sounds reasonable to me.

**Comment by conath** [comment]
@osy I've been looking into this, and it seems the 8 is only a hard limit with default QEMU settings (specifically, GICv2 - explainer see below). Here's the quote from the QEMU source that determines how many CPUs are supported for `virt`:

```c
    /* The maximum number of CPUs depends on the GIC version, or on how
     * many redistributors we can fit into the memory map.
     */
    if (vms->gic_version == VIRT_GIC_VERSION_3) {
        virt_max_cpus =
            vms->memmap[VIRT_GIC_REDIST].size / GICV3_REDIST_SIZE;
        virt_max_cpus +=
            vms->memmap[VIRT_HIGH_GIC_REDIST2].size / GICV3_REDIST_SIZE;
    } else {
        virt_max_cpus = GIC_NCPU;
    }

    if (max_cpus > virt_max_cpus) {
        error_report("Number of SMP CPUs requested (%d) exceeds max CPUs "
                     "supported by machine 'mach-virt' (%d)",
                     max_cpus, virt_max_cpus);
        exit(1);
    }
```
> QEMU Source `qemu/hw/arm/virt.c` [lines 1956-1970](https://github.com/qemu/qemu/blob/cc23377516072d293b0f9ab04741f8982d274639/hw/arm/virt.c#L1956)

GIC is the [Generic Interrupt Controller](https://developer.arm.com/architectures/system-architectures/system-components/arm-generic-interrupt-controller), available in QEMU in version 2 and 3. (QEMU `gic-version` parameter; defaults to 2).

GIC version 2 only supports max. of 8 CPUs:

```c
/* Maximum number of possible CPU interfaces, determined by GIC architecture */
#define GIC_NCPU 8
```
> QEMU source qemu/include/hw/intc/arm_gic_common.h [lines 32-33](https://github.com/qemu/qemu/blob/266469947161aa10b1d36843580d369d5aa38589/include/hw/intc/arm_gic_common.h#L33)

GIC version 3 has a dynamic CPU limit, see the first quote.

For testing, I added the `,gic_version=3` parameter to the QEMU Machine Properties in UTM 2.4.0 and started a `virt` machine I had previously created. It uses hvf, as I am testing on M1. The result was that the EFI BIOS seemed to hang at "Start boot option". It did not show the 3s timeout bar and pressing esc did not go to the EFI menu. Nothing appears in the Debug Log:

```log
Running:  -L /Applications/UTM/UTM.app/Contents/Resources/qemu -S -qmp tcp:127.0.0.1:4000,server,nowait -nodefaults -vga none -spice "unix=on,addr=/….spice,disable-ticketing=on,image-compression=off,playback-compression=off,streaming-video=off,gl=on" -device virtio-ramfb-gl -cpu host -smp cpus=2,sockets=1,cores=2,threads=1 -machine virt,highmem=off,gic_version=3 -accel hvf -accel tcg,tb-size=512 -boot menu=on -m 2048 -name "Ubuntu 20.04 ARM 2.1.2" -device qemu-xhci,id=usb-bus -device usb-tablet,bus=usb-bus.0 -device usb-mouse,bus=usb-bus.0 -device usb-kbd,bus=usb-bus.0 -device ich9-usb-ehci1,id=usb-controller-0 -device ich9-usb-uhci1,masterbus=usb-controller-0.0,firstport=0,multifunction=on -device ich9-usb-uhci2,masterbus=usb-controller-0.0,firstport=2,multifunction=on -device ich9-usb-uhci3,masterbus=usb-controller-0.0,firstport=4,multifunction=on -chardev spicevmc,name=usbredir,id=usbredirchardev0 -device usb-redir,chardev=usbredirchardev0,id=usbredirdev0,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev1 -device usb-redir,chardev=usbredirchardev1,id=usbredirdev1,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev2 -device usb-redir,chardev=usbredirchardev2,id=usbredirdev2,bus=usb-controller-0.0 -device usb-storage,drive=drive0,removable=true,bootindex=0,bus=usb-bus.0 -drive if=none,media=cdrom,id=drive0 -device virtio-blk-pci,drive=drive1,bootindex=1 -drive "if=none,media=disk,id=drive1,file=/…/disk-0.qcow2,cache=writethrough" -drive "if=pflash,media=disk,id=drive2,file=/…/edk2-aarch64-code.fd,cache=writethrough" -drive "if=pflash,media=disk,id=drive3,file=/…/edk2-arm-vars.fd,cache=writethrough" -device virtio-net-pci,mac=…,netdev=net0 -netdev user,id=net0 -uuid … -rtc base=localtime
WARNING: Image format was not specified for '/Users/chris/Library/Containers/com.utmapp.UTM/Data/Documents/Ubuntu 20.04 ARM 2.1.2.utm/Images/edk2-aarch64-code.fd' and probing guessed raw.
         Automatically detecting the format is dangerous for raw images, write operations on block 0 will be restricted.
         Specify the 'raw' format explicitly to remove the restrictions.
WARNING: Image format was not specified for '/Users/chris/Library/Containers/com.utmapp.UTM/Data/Documents/Ubuntu 20.04 ARM 2.1.2.utm/Images/edk2-arm-vars.fd' and probing guessed raw.
         Automatically detecting the format is dangerous for raw images, write operations on block 0 will be restricted.
         Specify the 'raw' format explicitly to remove the restrictions.
```

Note the missing lines `gl_version 30 - es profile enabled` and `WARNING: running without ARB/KHR robustness in place may crash` that would usually appear. To verify that our build of QEMU supports GICv3, I enabled "force slower emulation" and started the VM again:

```log
Running:  -L /Applications/UTM/UTM.app/Contents/Resources/qemu -S -qmp tcp:127.0.0.1:4000,server,nowait -nodefaults -vga none -spice "unix=on,addr=/….spice,disable-ticketing=on,image-compression=off,playback-compression=off,streaming-video=off,gl=on" -device virtio-ramfb-gl -cpu cortex-a57 -smp cpus=2,sockets=1,cores=2,threads=1 -machine virt,highmem=off,gic_version=3 -accel tcg,tb-size=512 -boot menu=on -m 2048 -name "Ubuntu 20.04 ARM 2.1.2" -device qemu-xhci,id=usb-bus -device usb-tablet,bus=usb-bus.0 -device usb-mouse,bus=usb-bus.0 -device usb-kbd,bus=usb-bus.0 -device ich9-usb-ehci1,id=usb-controller-0 -device ich9-usb-uhci1,masterbus=usb-controller-0.0,firstport=0,multifunction=on -device ich9-usb-uhci2,masterbus=usb-controller-0.0,firstport=2,multifunction=on -device ich9-usb-uhci3,masterbus=usb-controller-0.0,firstport=4,multifunction=on -chardev spicevmc,name=usbredir,id=usbredirchardev0 -device usb-redir,chardev=usbredirchardev0,id=usbredirdev0,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev1 -device usb-redir,chardev=usbredirchardev1,id=usbredirdev1,bus=usb-controller-0.0 -chardev spicevmc,name=usbredir,id=usbredirchardev2 -device usb-redir,chardev=usbredirchardev2,id=usbredirdev2,bus=usb-controller-0.0 -device usb-storage,drive=drive0,removable=true,bootindex=0,bus=usb-bus.0 -drive if=none,media=cdrom,id=drive0 -device virtio-blk-pci,drive=drive1,bootindex=1 -drive "if=none,media=disk,id=drive1,file=/…/disk-0.qcow2,cache=writethrough" -drive "if=pflash,media=disk,id=drive2,file=/…/edk2-aarch64-code.fd,cache=writethrough" -drive "if=pflash,media=disk,id=drive3,file=/…/edk2-arm-vars.fd,cache=writethrough" -device virtio-net-pci,mac=…,netdev=net0 -netdev user,id=net0 -uuid … -rtc base=localtime
WARNING: Image format was not specified for '/…/edk2-aarch64-code.fd' and probing guessed raw.
         Automatically detecting the format is dangerous for raw images, write operations on block 0 will be restricted.
         Specify the 'raw' format explicitly to remove the restrictions.
WARNING: Image format was not specified for '/…/edk2-arm-vars.fd' and probing guessed raw.
         Automatically detecting the format is dangerous for raw images, write operations on block 0 will be restricted.
         Specify the 'raw' format explicitly to remove the restrictions.
gl_version 30 - es profile enabled
WARNING: running without ARB/KHR robustness in place may crash
```

This time, the VM started up normally, just slower due to emulation instead of hvf virtualization. I tried again with 12 CPU cores, to verify we can fix this issue if we get GICv3 working with hvf. It worked fine, the only change in the log is the `smp cpus=12,sockets=1,cores=12,threads=1`. 

![Ubuntu VM running in UTM, showing the output of cpuinfo command. There are 12 cpu cores available in the guest.](https://user-images.githubusercontent.com/12073163/140318100-af4e4c8b-790a-423e-af3e-dcffe0e672a8.jpg)

In conclusion, it seems that (at least on my M1) with `hvf` mode and the included EFI BIOS, you can't boot a VM using QEMU GICv3. When you disable HVF, it works.

Machine: MacBook Pro M1
OS: 12.1 beta
UTM: 2.4.0 from GitHub

**Comment by ktprograms** [comment]
@conath the first search for `qemu gicv3 hvf` gets this: https://lore.kernel.org/qemu-devel/CAFEAcA8qwTY7OZzZSCqD3CA_wsL9VJEFWkCSNAerQohM4VvMuQ@mail.gmail.com/, and if you look at the end of the first block of `> `'s you can see: `- missing GICv3 support`

**Comment by osy** [comment]
Okay thanks for the find, I'll take out the 8 core on hvf code since that could change in the future with GICv3. But I'll keep in default to using pcores only which has the side effect of working on new M1 Macbooks which has 8 pcores.

**Comment by conath** [comment]
@osy do you want to apply the patch linked to by ktprograms? Then we could set GICv3 as default. That would actually solve the underlying issue which will otherwise appear again when Apple Silicon with more than 8 cores is introduced in the future.

**Comment by osy** [comment]
What do you mean? The patch says it's missing GICv3

**Comment by conath** [comment]
No, I mean [this one](https://lists.gnu.org/archive/html/qemu-devel/2021-01/msg05272.html).

**Comment by osy** [comment]
Since @agraf removed that patch in v7, I assume it's not ready.

**Comment by jandubois** [comment]
@agraf 

> I'd suggest to leave it at GICv2 for now until we sort out GICv3 support upstream.

Has there been any movement on this, or is it a lost cause (use Apple Virtualization Framework instead)?

I looked at the recent QEMU changelog, and see GICv3 mentioned quite a bit, but don't understand enough to be able to tell if this is related to the CPU count issue or not.

**Comment by ForbiddenEra** [comment]
> The patch works, but upstream is unhappy with the way it integrates in the rest of the code base, so I removed it to get anything upstream at all. Given that there are no Macs out with more than 8 p cores, I'd suggest to leave it at GICv2 for now until we sort out GICv3 support upstream.

16 pcores has been here for a while :)

**Comment by agraf** [comment]
The QEMU issue to track GICv3 support which would enable -smp > 8 is this: https://gitlab.com/qemu-project/qemu/-/issues/743. I would appreciate Tested-by / Reviewed-by tags on the mailing list to push it forward :).

**Comment by osy** [comment]
I'll open a new issue to track this so we don't forget.
